libs/medic-users-meta/migrations/202304140052.do.telemetry_devices.sql

```diff
@@ -9,26 +9,28 @@ SELECT
   doc #>> '{_id}' AS telemetry_doc_id,
   doc #>> '{metadata,deviceId}' AS device_id,
   doc #>> '{metadata,user}' AS user_name,
-  
-  concat_ws(
-    '-',
-    doc #>> '{metadata,year}',
-    CASE
-      WHEN 
-        doc #>> '{metadata,day}' IS NULL 
-        AND (
-          doc #>> '{metadata,versions,app}' IS NULL 
-          OR string_to_array("substring"(doc #>> '{metadata,versions,app}', '(\d+.\d+.\d+)'), '.')::integer[] < '{3,8,0}'::integer[]
-        ) 
-      THEN (doc #>> '{metadata,month}')::integer + 1
-      ELSE (doc #>> '{metadata,month}')::integer
-    END,
-    CASE
-      WHEN doc #>> '{metadata,day}' IS NOT NULL 
-      THEN doc #>> '{metadata,day}'
-      ELSE '1'
-    END
-  )::date AS period_start,
+  (CASE 
+      WHEN doc#>>'{metadata,year}' IS NULL THEN '1970-1-1'
```

Comment:
can this be null instead of random date?

Reply:
Hi @kennsippell , There'll be difference in results when using `NULL` vs `1970-1-1`. Since we have done `DISTINCT ON (doc #>> '{metadata,deviceId}', doc #>> '{metadata,user}')`, If we put NULL on our field, the NON-NULL value will be selected by default which in our case is first record and old record. If we put `1970-1-1`, it will consider the latest data aailable for those user and device. Since in our case, most of the records with `null` are recent, we will not be creating metadata for recent numbers. 

With this, there'll be two cases. 

1. Keep NULL, and get older records, most likely prior to 4.5.0 upgrades, period_start will have good older date and old data will be available. 
2. Keep 1970-1-1 and get recent record for that user and device. `period_start` will be 1970-1-1 in this case and latest data available. 

Let me know what'd you prefer here.

Reply:
IC. The `distinct on` should pick the first value ordered by `device_id, user_name, period_start`.  NULL is always sorted last by postgres. `1970` is likely also always going to be last, so they seem pretty close in functionality (?)

A third option is that we could use the 4.5.0 release date, since we know the telemetry doc is "at least that old" ? Nov-20-2023. Maybe that is getting too weird.

I think all three are decent options. You probably know best since you're deeper in the data atm. I'd likely go with NULL just to avoid "magic values" - but have no strong opinion and would respect either choice.

Reply:
Since we already have issue on cht-core to fix this problem of null telemetry dates, it doesn't really matter once that's fixed on what we've put here for useview_telemetry_devices. For other views, I opted to go with `NULL`.